### PR TITLE
[test/spec] Add test cases for unset, localvar and tempenv.

### PR DESCRIPTION
**Pull request by akinomyoga**
Test cases for #706

### CODE REVIEW

**Comment by andychu** [comment]
Wow this is very comprehensive.  I was thinking this file is more for idioms in real code, not synthetic test cases, but if you think it will come in handy it's OK.  

However I did notice some failures on ash?  Do you have this version?

```
andy@lisa:~/git/oilshell/oil2$ _deps/spec-bin/busybox 
BusyBox v1.31.1 (2020-04-18 02:00:51 PDT) multi-call binary.
BusyBox is copyrighted by many authors between 1998-2015.
Licensed under GPLv2. See source distribution for detailed
copyright notices.
```

The errors are in cases 14 and 15:

```
$ test/spec.sh ble-idioms -r 13-16
ble-idioms.test.sh
case    line    bash    zsh     mksh    ash     dash    yash    osh
 13     257     pass    ok      ok      ok      ok      pass    pass    [bash_unset] local-unset / dynamic-unset for tempenv
 14     290     BUG     ok      ok      FAIL    ok      ok      ok      [bash_unset] function call with tempenv vs tempenv-eval
 15     404     pass    ok      ok      FAIL    ok      ok      ok      [bash_unset] localvar-inherit from tempenv
 16     486     pass    pass    ok      ok      ok      ok      pass    [bash_unset] nested context by tempenv-eval

```

**Comment by andychu** [comment]
For case 14 it says this:


```
case: 14
[ash stdout] Expected '[global] v: global\n[global,local] v: (unset)\n[global,local+unset] v: (unset)\n[global,local+unlocal] v: (unset)\n[global,tempenv] v: tempenv\n[global,tempenv,local] v: tempenv\n[global,tempenv,local+unset] v: (uns
et)\n[global,tempenv,local+unlocal] v: (unset)\n[global,tempenv,(eval)] v: tempenv\n[global,tempenv,(eval),local] v: (unset)\n[global,tempenv,(eval),local+unset] v: (unset)\n[global,tempenv,(eval),local+unlocal] v: (unset)\n', got '[globa
l] v: global\n[global,local] v: global\n[global,local+unset] v: (unset)\n[global,local+unlocal] v: (unset)\n[global,tempenv] v: tempenv\n[global,tempenv,local] v: tempenv\n[global,tempenv,local+unset] v: (unset)\n[global,tempenv,local+unl
ocal] v: (unset)\n[global,tempenv,(eval)] v: tempenv\n[global,tempenv,(eval),local] v: tempenv\n[global,tempenv,(eval),local+unset] v: (unset)\n[global,tempenv,(eval),local+unlocal] v: (unset)\n'
ash stdout:
[global] v: global
[global,local] v: global
[global,local+unset] v: (unset)
[global,local+unlocal] v: (unset)
[global,tempenv] v: tempenv
[global,tempenv,local] v: tempenv
[global,tempenv,local+unset] v: (unset)
[global,tempenv,local+unlocal] v: (unset)
[global,tempenv,(eval)] v: tempenv
[global,tempenv,(eval),local] v: tempenv
[global,tempenv,(eval),local+unset] v: (unset)
[global,tempenv,(eval),local+unlocal] v: (unset)
```

**Comment by akinomyoga** [comment]
> However I did notice some failures on ash? Do you have this version?

Oh, really. I'm using the same version...

```bash
[murase@chatoyancy 1 oilshell.oil]$ _deps/spec-bin/busybox
BusyBox v1.31.1 (2020-03-19 18:41:37 JST) multi-call binary.
BusyBox is copyrighted by many authors between 1998-2015.
Licensed under GPLv2. See source distribution for detailed
copyright notices.

(snip)
```

The output of test case 14 from my ash:
```
[global] v: global
[global,local] v: (unset)
[global,local+unset] v: (unset)
[global,local+unlocal] v: (unset)
[global,tempenv] v: tempenv
[global,tempenv,local] v: tempenv
[global,tempenv,local+unset] v: (unset)
[global,tempenv,local+unlocal] v: (unset)
[global,tempenv,(eval)] v: tempenv
[global,tempenv,(eval),local] v: (unset)
[global,tempenv,(eval),local+unset] v: (unset)
[global,tempenv,(eval),local+unlocal] v: (unset)
```

The output of test case 15 from my ash:
```
[global,(local)] v: (unset)
[global,tempenv,(local)] v: tempenv
[xglobal,(local)] v: (unset)
[global,(func),(local)] v: (unset)
[global,tempenv,(func),(local)] v: (unset)
[xglobal,(func),(local)] v: (unset)
[global,local,(func),(local)] v: (unset)
```

Can you test the following command?
```bash
$ _deps/spec-bin/busybox sh -c 'fn() { local v; echo "v: ${v-(unset)}"; }; v=global; fn'
v: (unset)
```

My system is Fedora 30
```bash
[murase@chatoyancy 1 oilshell.oil]$ uname -a
Linux chatoyancy 5.1.20-300.fc30.x86_64 #1 SMP Fri Jul 26 15:03:11 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
[murase@chatoyancy 1 oilshell.oil]$ cat /etc/redhat-release
Fedora release 30 (Thirty)
```

**Comment by akinomyoga** [comment]
> I was thinking this file is more for idioms in real code, not synthetic test cases, but if you think it will come in handy it's OK.

OK! I created `spec/ble-features.test.sh` as suggested in https://github.com/oilshell/oil/issues/653#issuecomment-611111613 and moved these tests to that file!

---

I tried this command also with Linux busybox 1.28.3, Cygwin busybox 1.23.2, FreeBSD busybox 1.31.0 where the same result `v: (unset)` is reproduced.

```bash
$ busybox sh -c 'fn() { local v; echo "v: ${v-(unset)}"; }; v=global; fn'
v: (unset)
```

**Comment by andychu** [comment]
Ah OK there was a bug in the environment, leading me to run the wrong ash:

https://github.com/oilshell/oil/commit/4e0002e795d3ab89c22f68191cf24cb36d385ec5

Passes now:

http://travis-ci.oilshell.org/jobs/2020-04-18__19-07-12.wwz/_tmp/spec/ble-features.html

Thanks!

File: test/spec.sh [hunk 1]
```diff
@@ -939,4 +939,9 @@ ble-idioms() {
           $BASH $ZSH $MKSH $BUSYBOX_ASH $OSH_LIST "$@"
 }
 
+ble-features() {
+  sh-spec spec/ble-idioms.test.sh --osh-failures-allowed 0 \
```
Comment (akinomyoga) [inline test/spec.sh:943]: > Passes now:
>
> http://travis-ci.oilshell.org/jobs/2020-04-18__19-07-12.wwz/_tmp/spec/ble-features.html

I'm really sorry, but I noticed that I made a mistake in moving tests to ble-features. This part should have been

```diff
@@ -943,1,943,1 @@ ble-features() {
-  sh-spec spec/ble-idioms.test.sh --osh-failures-allowed 0 \
+  sh-spec spec/ble-features.test.sh --osh-failures-allowed 0 \
```

**Comment by andychu** [comment]
OK no problem!  I fixed it: https://github.com/oilshell/oil/commit/5753cba8b318dd52aecaa77ee608285dd0f2555e

**Comment by akinomyoga** [comment]
Thank you!

**Pull request merged**